katdal/test/test_lazy_indexer.py

```diff
@@ -73,19 +74,182 @@ def test_too_many_axes(self):
+    new_axes = tuple(n for n, k in enumerate(keep) if k is np.newaxis)
+    drop_axes = tuple(n for n, k in enumerate(keep) if _is_integer(k))
+    # Get rid of newaxis
+    keep = tuple(k for k in keep if k is not np.newaxis)
+    keep = ix_(keep, x.shape)
+    result = x[keep]
+    for ax in new_axes:
+        result = np.expand_dims(result, ax)
+    result = result.squeeze(axis=drop_axes)
+    return result
+
+
+def numpy_oindex_lite(x, keep):
+    """Perform outer indexing on a NumPy array (compact version).
+
+    This is more compact, but calls `x.__getitem__` `x.ndim` times.
+
+    It also assumes that `keep` contains no ellipsis to be as pure as possible.
+    """
+    if not isinstance(keep, tuple):
+        keep = (keep,)
+    dim = 0
+    result = x
+    for k in keep:
+        cumulative_index = (slice(None),) * dim + (k,)
+        result = result[cumulative_index]
+        # Handle dropped dimensions
+        if not _is_integer(k):
+            dim += 1
+    return result
+
+
+class TestNumPyOIndex(object):
+    """Test the :func:`~numpy_oindex` function."""
+    def setup(self):
+        shape = (10, 10, 10, 10)
+        self.data = np.arange(np.product(shape)).reshape(shape)
+
+    def _test_with(self, indices, indices_without_ellipsis=None):
+        """An even more compact version of numpy_oindex..."""
+        outer = numpy_oindex(self.data, indices)
+        if indices_without_ellipsis is None:
+            indices_without_ellipsis = indices
+        expected = numpy_oindex_lite(self.data, indices_without_ellipsis)
+        np.testing.assert_array_equal(outer, expected)
+
+    def test_outer_indexing(self):
+        self._test_with(())
+        self._test_with(2)
+        self._test_with((2, 3, 4, 5))
+        # ellipsis
+        self._test_with(np.s_[[0], ...], np.s_[[0], :, :, :])
+        self._test_with(np.s_[:, [0], ...], np.s_[:, [0], :, :])
+        # list of ints
+        self._test_with(np.s_[:, [0], [0], :])
+        self._test_with(np.s_[:, [0], :, [0]])
+        self._test_with(np.s_[:, [0], [0, 1], :])
+        self._test_with(np.s_[:, [0], :, [0, 1]])
+        # booleans
+        pick_one = np.zeros(10, dtype=np.bool_)
+        pick_one[6] = True
+        self._test_with(np.s_[:, [True, False] * 5, pick_one, :])
+        self._test_with(np.s_[:, [False, True] * 5, :, pick_one])
+        # slices
+        self._test_with(np.s_[0:2, 2:4, 4:6, 6:8])
+        self._test_with(np.s_[-8:-6, -4:-2, slice(3, 10, 2), -2:])
+        # single ints
+        self._test_with(np.s_[:, [0], 0, :])
+        self._test_with(np.s_[:, [0], :, 0])
+        self._test_with(np.s_[:, [0], -1, :])
+        self._test_with(np.s_[:, [0], :, -1])
+        # newaxis
+        self._test_with(np.s_[np.newaxis, :, [0], :, 0])
+        self._test_with(np.s_[:, [0], np.newaxis, 0, :])
+        # the lot
+        self._test_with(np.s_[..., 2:5, [4, 6], np.newaxis, 0],
+                        np.s_[:, 2:5, [4, 6], np.newaxis, 0])
+
```

Comment:
Could you share the list of test cases between this test and TestDaskGetitem? Maybe ellipsis cases might need some special handling, but it looks like there is a lot of duplication.

Reply:
Done.